### PR TITLE
feat: Support validating signed payload

### PR DESCRIPTION
**Pull request by armenzg**
This allows setting a secret in the Github webhook and validating.

### CODE REVIEW

File: src/event_handler.py [hunk 1]
```diff
@@ -1,20 +1,41 @@
...
     def handle_event(self, data, headers):
         # We return 200 to make webhook not turn red since everything got processed well
         http_code = 200
         reason = "OK"
 
-        if headers["X-GitHub-Event"] != "workflow_job":
+        if self.secret and not valid_payload(
+            self.secret, data, headers["X-Hub-Signature"].replace("sha1=", "")
+        ):
+            http_code = 400
```
Comment (armenzg) [inline src/event_handler.py:25]: If a developer set ups their Github webhook with a bad secret, the UI for delivered payloads will turn red for each event, thus, making it easy for them to realize that their set up secret is invalid.

File: tests/fixtures/webhook_event.json [hunk 2]
```diff
@@ -1,72 +1,173 @@
 {
-  "action": "completed",
```
Comment (armenzg) [inline tests/fixtures/webhook_event.json:2]: This body has been moved within a top-level `payload` key.

File: src/event_handler.py [hunk 3]
```diff
@@ -14,7 +19,23 @@ def handle_event(self, data, headers):
             reason = "Event not supported."
         elif data["action"] != "completed":
             reason = "We cannot do anything with this workflow state."
+        elif self.secret and not valid_payload(
+            self.secret, data, headers["X-Hub-Signature"].replace("sha1=", "")
```
Comment (asottile-sentry) [inline src/event_handler.py:23]: we should use the new X-Hub-Signature-256

File: src/event_handler.py [hunk 4]
```diff
@@ -14,7 +19,23 @@ def handle_event(self, data, headers):
...
+            reason = "The secret you are using on your Github webhook does not match this app's secret."
         else:
             self.client.send_trace(data["workflow_job"])
 
         return reason, http_code
+
+
+def payload_signature(secret, payload):
+    return hmac.new(
+        secret.encode("utf-8"), json.dumps(payload).encode("utf-8"), hashlib.sha1
```
Comment (asottile-sentry) [inline src/event_handler.py:35]: fwiw, the default encoding is utf-8 so you can just do `secret.encode()`

Comment (asottile-sentry) [inline src/event_handler.py:35]: this won't necessarily match properly (loading and then dumping json could have ~slightly different encodings depending on how github chooses to encode (json has several ascii-vs-non-ascii escaping flavors)) -- github signs the exact bytes of the input so we should probably do that as well.  I use something like this in pre-commit.ci:

```python
def _github_validate_signature(request: flask.Request) -> None:
    sig = hmac.new(creds.github_webhook_secret(), digestmod='sha256')
    sig.update(request.data)
    signature = f'sha256={sig.hexdigest()}'

    if 'X-Hub-Signature-256' not in request.headers:
        flask.abort(400, description='missing X-Hub-Signature-256')

    hub_signature = request.headers['X-Hub-Signature-256']
    if not hmac.compare_digest(signature, hub_signature):
        flask.abort(400, description='invalid signature')
```

File: tests/test_handle_event.py [hunk 5]
```diff
@@ -1,12 +1,15 @@
 import pytest
-from src.event_handler import EventHandler
+
+from src.event_handler import EventHandler, valid_payload
+from .fixtures import *
```
Comment (asottile-sentry) [inline tests/test_handle_event.py:4]: eheh we should set up some linters -- or move the fixtures to `conftest.py`.  making fixtures available by import side-effects is ~technically not a supported mechanism in pytest (they should either be in a `conftest` or a pytest plugin) and only kinda work by accident right now

Comment (armenzg) [reply tests/test_handle_event.py:4]: Isn't it more fun to walk off the beaten path? :) I will look into it

**Comment by armenzg** [comment]
I will handle importing the fixtures in a follow up PR.

**Comment by armenzg** [comment]
@asottile-sentry is this good to go?

**Comment by asottile-sentry** [comment]
> @asottile-sentry is this good to go?

oh, I was hoping you'd do the 256 stuff first -- that way we're not using broken crypto

**Comment by asottile-sentry** [comment]
> > @asottile-sentry is this good to go?
> 
> oh, I was hoping you'd do the 256 stuff first -- that way we're not using broken crypto

plus the signing of the actual request bytes rather than the normalized json

**Comment by armenzg** [comment]
> > > @asottile-sentry is this good to go?
> > 
> > 
> > oh, I was hoping you'd do the 256 stuff first -- that way we're not using broken crypto
> 
> plus the signing of the actual request bytes rather than the normalized json

My apologies. I only saw part of all your comments. I will have it ready by tomorrow.

File: src/event_handler.py [hunk 6]
```diff
@@ -32,7 +30,7 @@ def handle_event(self, data, headers):
 
 def payload_signature(secret, payload):
     return hmac.new(
-        secret.encode("utf-8"), json.dumps(payload).encode("utf-8"), hashlib.sha1
+        secret.encode(), msg=str(payload).encode(), digestmod="sha256"
```
Comment (armenzg) [inline src/event_handler.py:33]: Is this better?

File: tests/test_github_sdk.py [hunk 7]
```diff
@@ -20,23 +20,34 @@ def test_job_without_steps(skipped_workflow):
 
 
 def test_initialize_without_setting_dsn():
-    with pytest.raises(TypeError):
+    with pytest.raises(TypeError) as excinfo:
```
Comment (armenzg) [inline tests/test_github_sdk.py:23]: I've followed the pattern into this test file as well.

**Pull request merged**